---
title: "scIGD on Merkel Cell Carcinoma"
author: "Ahmad Al Ajami"
date: "`r Sys.Date()`"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 5
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, include=TRUE, message=FALSE, cache = FALSE}
suppressPackageStartupMessages({
library(SingleCellAlleleExperiment)
library(tidyverse)
library(gridExtra)
library(DropletUtils)
library(BiocParallel)
library(scDblFinder)
library(Matrix)
library(AnnotationHub)
library(cowplot)
library(pheatmap)
})
```

# SingleCellAlleleExperiment

```{r scae, message=FALSE}
donor_names <- list.dirs("data/", full.names = FALSE)[-1] %>% 
  grep("gen", ., value = TRUE)

read_scae <- function(folder_path){
  folder_name <- gsub("data/", "", folder_path)
  counts <- read_allele_counts(folder_path,
                               lookup_file = read.csv(file.path(folder_path, "lookup_table_HLA.csv")),
                               sample_names = folder_name,
                               filter_mode = "yes",
                               log = TRUE,
                               gene_symbols = TRUE)
  return(counts)
}

scae_list <- lapply(c("pbmc_postRx", "pbmc_preRx", "tumor_postRx", "tumor_preRx"), function(donor){
  read_scae(file.path("data", donor)) 
})

# work around for tumor_postRx
scae_list[[3]] <- scae_list[[3]][intersect(rownames(scae_list[[1]]), rownames(scae_list[[3]])), ]
```

## Adding metadata

The metadata contain information regarding the tissue and timepoint

```{r}
scae_list <- lapply(seq_along(scae_list), function(i) {
  scae <- scae_list[[i]]
  colData(scae)$state <- donor_names[i]
  return(scae)
})
```

## Data summary: pre-processing

```{r}
dt <- lapply(scae_list, function(scae) {
    data.frame(State = unique(scae$state),
               Number_of_features = nrow(scae),
               Number_of_cells = ncol(scae))
})
dt <- do.call(rbind, dt)

DT::datatable(dt, 
              class = 'compact stripe hower',
              options = list(
                dom = "Bfrtip",
                scrollX = TRUE,
                paging = TRUE,
                searching = TRUE,
                info = TRUE,
                ordering = TRUE,
                columnDefs = list(list(className = 'dt-center', targets = 0:2))), rownames = FALSE)
```

Which alleles are present in each gene:

```{r, fig.height=4, fig.width=12}
df <- lapply(scae_list, function(sce){
  data.frame(allele = scae_subset(sce, subset = "alleles") %>% rownames) %>%
    mutate(gene = paste0("HLA-", str_extract(allele, "^[^*]+")))
}) %>% 
  bind_rows() %>% 
  group_by(gene, allele) %>%
  mutate(count = n()) %>% distinct(gene, allele, .keep_all = TRUE)

p <- by(data = df, INDICES = df$gene, FUN = function(m) {
    m <- droplevels(m)
    m <- ggplot(m, aes(x = gene, y = count, fill = allele)) +
        geom_bar(stat = "identity") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        facet_wrap(~gene) +
        theme(legend.position = "bottom")
})

do.call(grid.arrange, c(p, ncol=3))
```

```{r, fig.height=14, fig.width=22}
df <- lapply(scae_list, function(sce){
  data.frame(state = unique(sce$state),
             allele = scae_subset(sce, subset = "alleles") %>% rownames) %>%
    mutate(gene = paste0("HLA-", str_extract(allele, "^[^*]+")))
}) %>% 
  bind_rows() %>% 
  group_by(allele, state) %>%
  mutate(count = n()) %>% distinct(allele, state, .keep_all = TRUE)

p <- by(data = df, INDICES = df$gene, FUN = function(m) {
    m <- droplevels(m)
    m <- ggplot(m, aes(x = allele, y = count, fill = state)) +
        geom_bar(stat = "identity") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        facet_wrap(~gene) +
        theme(legend.position = "bottom")
})

do.call(grid.arrange, c(p, ncol=1))
```

# Pre-processing workflow

## Looking at distributions

Plotting the normalized expression of alleles:

```{r, allele_expression}
plot_feature_expression <- function(scae, scae_feature) {
  plotExpression(scae, scae_feature) +
    labs(title = scae$state) +
    theme_bw() +
    theme(axis.text = element_text(size = 5),
          axis.title = element_text(size = 8, face = "bold"), 
          axis.text.x = element_text(angle = 45, hjust = 1))
}

lapply(scae_list, function(scae) {
  scae_feature <- c(scae_subset(scae, subset = "alleles") %>% rownames,
                scae_subset(scae, subset = "immune_genes") %>% rownames,
                scae_subset(scae, subset = "functional_groups") %>% rownames)
  print(paste("number of cells:", ncol(scae)))
  plot_feature_expression(scae, scae_feature)
})
```

```{r}
plot_total_counts <- function(scae) {
  total_counts <- colSums(counts(scae))
  
  ggplot(data.frame(Total_Counts = total_counts), aes(x = Total_Counts)) +
    geom_histogram(binwidth = 500, fill = "skyblue", color = "black") +
    labs(title = scae$state,
         x = "total counts per cell",
         y = "number of cells") +
    theme_bw()
}

lapply(scae_list, plot_total_counts)
```

## Cell filtering

### Removing within-sample doublets

```{r doublets, message=FALSE, warning=FALSE}
set.seed(42L)

doublet_finder <- function(scae){
  scae <- scDblFinder(scae, BPPARAM=MulticoreParam(4))
  tbl <- table(scae$scDblFinder.class)
  cat(unique(scae$state), "\n", paste(names(tbl), tbl, sep = " "), "\n")
  scae <- scae[, scae$scDblFinder.class == "singlet"]
  return(scae)
}

scae_list <- lapply(scae_list, doublet_finder)
```

### Removing low quality cells

```{r}
scae_list <- lapply(scae_list, function(scae){
  rownames(scae) <- sub("\\..*", "", rownames(scae))
  return(scae)
})

# identifying the mitochondrial transcripts:
query(AnnotationHub(), "EnsDb.Hsapiens.")

# release 93 from cellranger
edb <- AnnotationHub()[["AH64446"]]
```

Some QC:

```{r, warning=FALSE}
qc_annotationhub <- function(scae){
  location <- mapIds(edb, keys=rownames(scae), keytype="GENEID", column="SEQNAME")
  is.mito <- which(location=="MT")
  df <- perCellQCMetrics(scae, subsets=list(Mito=is.mito))
  reasons <- quickPerCellQC(df, percent_subsets=c("subsets_Mito_percent"))
  tbl <- colSums(as.matrix(reasons))
  cat(unique(scae$state), "\n", paste(names(tbl), tbl, sep = " "), "\n")
  colData(scae) <- cbind(colData(scae), df)
  scae$discard <- reasons$discard
  return(scae)
}

scae_list <- lapply(scae_list, qc_annotationhub)
```

```{r, fig.width=10, fig.height=4}
qc_plots <- function(scae){
  gridExtra::grid.arrange(
    plotColData(scae, x="state", y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(scae, x="state", y="detected", colour_by="discard") + 
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(scae, x="state", y="subsets_Mito_percent", colour_by="discard") + 
      ggtitle("Mito percent"),
    ncol=3)
}

lapply(scae_list, qc_plots) %>% print
```

```{r}
qc_mito_plot <- function(scae){
  plotColData(scae, x="sum", y="subsets_Mito_percent", colour_by="discard") +
    theme(panel.border = element_rect(color = "grey")) + 
    ggtitle(unique(scae$Donor_ID))
}

lapply(scae_list, qc_mito_plot) %>% print()
```

Now applying filter and removing "discarded" cells

```{r}
apply_qc_filter <- function(scae){
  scae <- scae[, !scae$discard]
  return(scae)
}

scae_list <- lapply(scae_list, apply_qc_filter)
```

## Feature filtering

Removing all features that have a total count less than 5; with the exception of all alleles - all alleles are retained

```{r}
'%nin%' <- Negate('%in%')

filter_features <- function(scae){
  alleles <- grep("^ENS", rownames(scae), invert=TRUE, value=TRUE)
  other_genes <- rownames(scae) %nin% alleles
    features_keep <- rowSums(counts(scae)[other_genes, ]) > 5
  features_keep_names <- names(features_keep)[which(features_keep)]
  scae <- scae[c(features_keep_names, alleles), ]
  return(scae)
}

scae_list <- lapply(scae_list, filter_features)
```

## After filtering

```{r}
dt <- lapply(scae_list, function(scae) {
    data.frame(State = unique(scae$state),
               Number_of_features = nrow(scae),
               Number_of_cells = ncol(scae))
})
dt <- do.call(rbind, dt)

DT::datatable(dt, 
              class = 'compact stripe hower',
              options = list(
                dom = "Bfrtip",
                scrollX = TRUE,
                paging = FALSE,
                searching = TRUE,
                info = TRUE,
                ordering = TRUE,
                columnDefs = list(list(className = 'dt-center', targets = 0:2))), rownames = FALSE)
```

# Processing workflow

## Normalization

The data is normalized upon creating the *SingleCellAlleleExperiment* object

## Variance modeling

```{r, warning=FALSE}
variance_modeling <- function(scae){
  gene_var <- modelGeneVar(scae)
  plot(gene_var$mean, gene_var$total, main=unique(scae$state), pch=16, cex=0.5,
       xlab="mean of log-expression", ylab="variance of log-expression")
  curfit <- metadata(gene_var)
  points(curfit$mean, curfit$var, col="red", pch=16)
  curve(curfit$trend(x), col='dodgerblue', add=TRUE, lwd=2) 
}

lapply(scae_list, variance_modeling) %>% print()
```

## Distribution of allele features

```{r, fig.asp=1}
plot_allele_dist <- function(scae){
  alleles <- grep("^ENS", rownames(scae), invert=TRUE, value=TRUE)
  p <- do.call(plot_grid, c(lapply(alleles, function(feature){
    ggplot(as.data.frame(logcounts(scae)[feature, ]), aes(x=logcounts(scae)[feature, ])) +
      geom_histogram(color="black",
                     fill = "black",
                     breaks=seq(0, 6, by=0.10)) +
      labs(title = unique(scae$state),
           x = feature) +
      theme(axis.title.x = element_text(size = 8),
            axis.title.y = element_text(size = 8),
            plot.title = element_text(size = 8))}), ncol = 4))
  print(p)
  cat("\n")
}

lapply(scae_list, plot_allele_dist)
```

```{r}
scae_list <- lapply(scae_list, function(scae){
  as(scae, "SingleCellAlleleExperiment")
})

lapply(scae_list, function(scae) {
  scae_feature <- c(scae_subset(scae, subset = "alleles") %>% rownames,
                scae_subset(scae, subset = "immune_genes") %>% rownames,
                scae_subset(scae, subset = "functional_groups") %>% rownames)
  plot_feature_expression(scae, scae_feature)
})
```

```{r, eval=FALSE}
saveRDS(scae_list, "scae.Rds")
```

# Session

```{r sessionInfo, cache = 0}
sessionInfo()
```

