---
title: "scIGD on AIDA"
author: "Ahmad Al Ajami"
date: "`r Sys.Date()`"
output:
  html_document:
    keep_md: false
    toc: true
    toc_float: true
    toc_depth: 5
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, include=TRUE, message=FALSE, cache = FALSE}
suppressPackageStartupMessages({
library(SingleCellAlleleExperiment)
library(tidyverse)
library(gridExtra)
library(DropletUtils)
library(BiocParallel)
library(scDblFinder)
library(Matrix)
library(AnnotationHub)
library(cowplot)
})
```

# Loading data

```{r}
raw_data <- read.csv("data/AIDApilotdata.tsv", header = TRUE, sep = "\t")
donor_names <- list.dirs("data/", full.names = FALSE)[-1]
data <- raw_data %>%
  dplyr::filter(bundle_uuid %in% donor_names) %>%
  dplyr::distinct(bundle_uuid, .keep_all = TRUE)

data_summary <- data[, c("bundle_uuid", "donor_organism.organism_age", "donor_organism.sex", "file_name")] %>% as.data.frame()
data_summary$donor_ID <- gsub("\\..*", "", data_summary$file_name)
data_summary$donor_ID <- gsub("_B.*?_5GEX", "", data_summary$donor_ID)
colnames(data_summary) <- c("Sample", "Age", "Sex", "Nationality", "Donor_ID")
data_summary$Nationality <- substr(data_summary$Nationality, 1, 2)
data_summary <- data_summary %>% 
  arrange(Age)


data_summary
```

# SingleCellAlleleExperiment

```{r scae, message=FALSE}
donor_names <- list.dirs("data/", full.names = FALSE)[-1]

read_scae <- function(folder_path){
  folder_name <- gsub("data/", "", folder_path)
  counts <- read_allele_counts(folder_path,
                               lookup_file = read.csv(file.path(folder_path, "lookup_table_HLA.csv")),
                               sample_names = folder_name,
                               filter_mode = "yes",
                               log = TRUE,
                               gene_symbols = TRUE)
  return(counts)
}

scae_list <- lapply(donor_names, function(donor){
  read_scae(file.path("data", donor)) 
})

sapply(scae_list, function(scae) ncol(scae))
```

## Adding metadata

The metadata contain information regarding age, sex, nationality and donor ID

```{r}
add_metadata <- function(scae){
    matching_rows <- data_summary$Sample %in% colData(scae)$Sample
    colData(scae) <- cbind(colData(scae), data_summary[matching_rows, -1])
    return(scae)
}

scae_list <- lapply(scae_list, add_metadata)
```

## Data summary: pre-processing

```{r}
dt <- lapply(scae_list, function(scae) {
    data.frame(Sample = unique(scae$Sample),
               Age = unique(scae$Age),
               Sex = unique(scae$Sex),
               Nationality = unique(scae$Nationality),
               Donor_ID = unique(scae$Donor_ID),
               Number_of_features = nrow(scae),
               Number_of_cells = ncol(scae))
})
dt <- do.call(rbind, dt)

DT::datatable(dt, 
              class = 'compact stripe hower',
              options = list(
                dom = "Bfrtip",
                scrollX = TRUE,
                paging = TRUE,
                searching = TRUE,
                info = TRUE,
                ordering = TRUE,
                columnDefs = list(list(className = 'dt-center', targets = 0:6))), rownames = FALSE)
```

- y-axis represents the number of cells (cell_count)
- x-axis represents nationality, sex and age, respectively
- fill represents donor ID

```{r, fig.height=14, fig.width=10}
get_metadata <- function(sce) {
  data.frame(donor_ID = unique(sce$Donor_ID),
             nationality = unique(sce$Nationality),
             age = unique(sce$Age) %>% gsub(" year", "", .),
             sex = unique(sce$Sex),
             cell_count = ncol(sce))
}
df <- lapply(scae_list, get_metadata) %>% 
  bind_rows()

p1 <- ggplot(df, aes(x = nationality, y = cell_count, fill = donor_ID)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none")

p2 <- ggplot(df, aes(x = sex, y = cell_count, fill = donor_ID)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "none")

p3 <- ggplot(df, aes(x = age, y = cell_count, fill = donor_ID)) +
  geom_bar(stat = "identity") +
  theme(legend.position = "bottom")

grid.arrange(p1, p2, p3, ncol = 1)
```

Which alleles are present in each gene; What is the occurrence of each of these alleles in the 14 donors we have, in other words, how many times is one allele repeated across the 14 donors we have:

```{r, fig.height=32, fig.width=10}
df <- lapply(scae_list, function(sce){
  data.frame(allele = scae_subset(sce, subset = "alleles") %>% rownames) %>%
    mutate(gene = paste0("HLA-", str_extract(allele, "^[^*]+")))
}) %>% 
  bind_rows() %>% 
  group_by(gene, allele) %>%
  mutate(count = n()) %>% distinct(gene, allele, .keep_all = TRUE)

p <- by(data = df, INDICES = df$gene, FUN = function(m) {
    m <- droplevels(m)
    m <- ggplot(m, aes(x = gene, y = count, fill = allele)) +
        geom_bar(stat = "identity") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        facet_wrap(~gene) +
        theme(legend.position = "bottom")
})

do.call(grid.arrange, c(p, ncol=1))
```

How many donors of Korean and Japanese origin are present in each of the alleles

y-axis reflects the occurrence of each of these alleles in the 14 donors we have:

```{r, fig.height=32, fig.width=22}
df <- lapply(scae_list, function(sce){
  data.frame(nationality = unique(sce$Nationality),
             allele = scae_subset(sce, subset = "alleles") %>% rownames) %>%
    mutate(gene = paste0("HLA-", str_extract(allele, "^[^*]+")))
}) %>% 
  bind_rows() %>% 
  group_by(allele, nationality) %>%
  mutate(count = n()) %>% distinct(allele, nationality, .keep_all = TRUE)

p <- by(data = df, INDICES = df$gene, FUN = function(m) {
    m <- droplevels(m)
    m <- ggplot(m, aes(x = allele, y = count, fill = nationality)) +
        geom_bar(stat = "identity") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        facet_wrap(~gene) +
        theme(legend.position = "bottom")
})

do.call(grid.arrange, c(p, ncol=1))
```

Which donors are present in each of the alleles; the count is equal to 1 for all donors

y-axis reflects the occurrence of each of these alleles in the 14 donors we have:

```{r, fig.height=42, fig.width=22}
df <- lapply(scae_list, function(sce){
  data.frame(donor_ID = unique(sce$Donor_ID),
             allele = scae_subset(sce, subset = "alleles") %>% rownames) %>%
    mutate(gene = paste0("HLA-", str_extract(allele, "^[^*]+")))
}) %>% 
  bind_rows() %>% 
  group_by(allele, donor_ID) %>%
  mutate(count = n()) %>% distinct(allele, donor_ID, .keep_all = TRUE)

p <- by(data = df, INDICES = df$gene, FUN = function(m) {
    m <- droplevels(m)
    m <- ggplot(m, aes(x = allele, y = count, fill = donor_ID)) +
        geom_bar(stat = "identity") +
        theme_bw() +
        theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
        facet_wrap(~gene) +
        theme(legend.position = "bottom")
})

do.call(grid.arrange, c(p, ncol=1))
```

# Pre-processing workflow

## Looking at distributions before filtering

```{r}
plot_total_counts <- function(scae) {
  total_counts <- colSums(counts(scae))
  
  ggplot(data.frame(Total_Counts = total_counts), aes(x = Total_Counts)) +
    geom_histogram(binwidth = 500, fill = "skyblue", color = "black") +
    labs(title = scae$Donor_ID,
         x = "total counts per cell",
         y = "number of cells") +
    theme_bw()
}

lapply(scae_list, plot_total_counts)
```

## Cell filtering

### Removing within-sample doublets

```{r doublets, message=FALSE, warning=FALSE}
set.seed(42L)

doublet_finder <- function(scae){
  scae <- scDblFinder(scae, BPPARAM=MulticoreParam(4))
  tbl <- table(scae$scDblFinder.class)
  cat(unique(scae$Donor_ID), "\n", paste(names(tbl), tbl, sep = " "), "\n")
  scae <- scae[, scae$scDblFinder.class == "singlet"]
  return(scae)
}

scae_list <- lapply(scae_list, doublet_finder)
```

### Removing low quality cells

```{r}
scae_list <- lapply(scae_list, function(scae){
  rownames(scae) <- sub("\\..*", "", rownames(scae))
  return(scae)
})

# identifying the mitochondrial transcripts:
query(AnnotationHub(), "EnsDb.Hsapiens.")

# release 93 from cellranger
edb <- AnnotationHub()[["AH64446"]]
```

Some QC:

```{r, warning=FALSE}
qc_annotationhub <- function(scae){
  location <- mapIds(edb, keys=rownames(scae), keytype="GENEID", column="SEQNAME")
  is.mito <- which(location=="MT")
  df <- perCellQCMetrics(scae, subsets=list(Mito=is.mito))
  reasons <- quickPerCellQC(df, percent_subsets=c("subsets_Mito_percent"))
  tbl <- colSums(as.matrix(reasons))
  cat(unique(scae$Donor_ID), "\n", paste(names(tbl), tbl, sep = " "), "\n")
  colData(scae) <- cbind(colData(scae), df)
  scae$discard <- reasons$discard
  return(scae)
}

scae_list <- lapply(scae_list, qc_annotationhub)
```

```{r, fig.width=10, fig.height=4}
qc_plots <- function(scae){
  gridExtra::grid.arrange(
    plotColData(scae, x="Donor_ID", y="sum", colour_by="discard") +
        scale_y_log10() + ggtitle("Total count"),
    plotColData(scae, x="Donor_ID", y="detected", colour_by="discard") + 
        scale_y_log10() + ggtitle("Detected features"),
    plotColData(scae, x="Donor_ID", y="subsets_Mito_percent",colour_by="discard") + 
      ggtitle("Mito percent"),
    ncol=3)
}

lapply(scae_list, qc_plots) %>% print
```

```{r}
qc_mito_plot <- function(scae){
  plotColData(scae, x="sum", y="subsets_Mito_percent", colour_by="discard") +
    theme(panel.border = element_rect(color = "grey")) + 
    ggtitle(unique(scae$Donor_ID))
}

lapply(scae_list, qc_mito_plot) %>% print()
```

Now applying filter and removing "discarded" cells

```{r}
apply_qc_filter <- function(scae){
  scae <- scae[, !scae$discard]
  return(scae)
}

scae_list <- lapply(scae_list, apply_qc_filter)
```

## Feature filtering

Removing all features that have a total count less than 5; with the exception of all alleles - all alleles are retained

```{r}
'%nin%' <- Negate('%in%')

filter_features <- function(scae){
  alleles <- grep("^ENS", rownames(scae), invert=TRUE, value=TRUE)
  other_genes <- rownames(scae) %nin% alleles
  
  features_keep <- rowSums(counts(scae)[other_genes, ]) > 5
  features_keep_names <- names(features_keep)[which(features_keep)]
  scae <- scae[c(features_keep_names, alleles), ]
  return(scae)
}

scae_list <- lapply(scae_list, filter_features)
```

## After filtering

```{r}
dt <- lapply(scae_list, function(scae) {
    data.frame(Sample = unique(scae$Sample),
               Age = unique(scae$Age),
               Sex = unique(scae$Sex),
               Nationality = unique(scae$Nationality),
               Donor_ID = unique(scae$Donor_ID),
               Number_of_features = nrow(scae),
               Number_of_cells = ncol(scae))
})
dt <- do.call(rbind, dt)

DT::datatable(dt, 
              class = 'compact stripe hower',
              options = list(
                dom = "Bfrtip",
                scrollX = TRUE,
                paging = FALSE,
                searching = TRUE,
                info = TRUE,
                ordering = TRUE,
                columnDefs = list(list(className = 'dt-center', targets = 0:6))), rownames = FALSE)
```

# Processing workflow

## Normalization

The data is normalized upon creating the *SingleCellAlleleExperiment* object

## Variance modeling

```{r, warning=FALSE}
variance_modeling <- function(scae){
  gene_var <- modelGeneVar(scae)
  plot(gene_var$mean, gene_var$total, main=unique(scae$Donor_ID), pch=16, cex=0.5,
       xlab="mean of log-expression", ylab="variance of log-expression")
  curfit <- metadata(gene_var)
  points(curfit$mean, curfit$var, col="red", pch=16)
  curve(curfit$trend(x), col='dodgerblue', add=TRUE, lwd=2) 
}

lapply(scae_list, variance_modeling) %>% print()
```

## Distribution of immune features

```{r, fig.asp=1.8}
plot_allele_dist <- function(scae){
  immune_features <- rownames(scae[rowData(scae)$NI_I == "I", ])  
  p <- do.call(plot_grid, c(lapply(immune_features, function(feature){
    ggplot(as.data.frame(logcounts(scae)[feature, ]), aes(x=logcounts(scae)[feature, ])) +
      geom_histogram(color="black",
                     fill = "black",
                     breaks=seq(0, 3, by=0.10)) +
      labs(title = unique(scae$Donor_ID),
           x = feature) +
      theme(axis.title.x = element_text(size = 8),
            axis.title.y = element_text(size = 8),
            plot.title = element_text(size = 8))}), ncol = 4))
  print(p)
  cat("\n")
}

lapply(scae_list, plot_allele_dist)
```

Plotting the normalized expression of alleles:

```{r, allele_expression}
plot_feature_expression <- function(scae, immune_features) {
  plotExpression(scae, immune_features) +
    labs(title = scae$Donor_ID) +
    theme_bw() +
    theme(axis.text = element_text(size = 5),
          axis.title = element_text(size = 8, face = "bold"), 
          axis.text.x = element_text(angle = 45, hjust = 1))
}

lapply(scae_list, function(scae) {
  immune_features <- rownames(scae[rowData(scae)$NI_I == "I", ])
  plot_feature_expression(scae, immune_features)
})
```

```{r, eval=FALSE}
saveRDS(scae_list, "scae.Rds")
```

# Session info

```{r sessionInfo, cache = 0}
sessionInfo()
```

